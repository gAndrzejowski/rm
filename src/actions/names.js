export const GET_MOVIES = 'get movies';
export const SORT_MOVIES = 'sort movies';
export const SET_CURRENT_MOVIE = 'set current movie';
export const SET_SEARCH_CRITERION = 'set search criterion';
export const SET_SEARCH_TXT = 'set search txt';
export const HYDRATE_STORE = 'hydrate store';

export default {
    GET_MOVIES,
    SORT_MOVIES,
    SET_SEARCH_CRITERION,
    SET_CURRENT_MOVIE,
    SET_SEARCH_TXT,
    HYDRATE_STORE
};