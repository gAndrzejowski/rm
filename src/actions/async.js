import swagger, {single} from '../util/calls';
import {get_movies, set_current_movie, sort_movies} from "./creators";

export const searchMovies = async (query, searchBy) => {
    const results = await swagger({
        search: query,
        searchBy
    });
    return get_movies(results.data);
};

export const searchAndSort = async (sortBy, search, searchBy) => {
    const results = await swagger({
        sortBy,
        search,
        searchBy
    });
    console.log(results.data.length);
    return sort_movies(sortBy, results.data);
};

export const getByGenre = async (genre) => {
    const results = await swagger({
        search: genre,
        searchBy: "genres"
    });
    return get_movies(results.data);
};

export const getById = async id => {
    const results = await single(id);
    return set_current_movie(results);
};

export default searchMovies;