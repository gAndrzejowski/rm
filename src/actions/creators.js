import actions from './names';

const {
    GET_MOVIES,
    SORT_MOVIES,
    SET_SEARCH_CRITERION,
    SET_CURRENT_MOVIE,
    SET_SEARCH_TXT,
    HYDRATE_STORE
} = actions;

export const get_movies = (payload) => ({type: GET_MOVIES, payload});
export const sort_movies = (by, results) => ({type: SORT_MOVIES, by, results});
export const set_search_criterion = (criterion) => ({type: SET_SEARCH_CRITERION, criterion});
export const set_current_movie = (movie) => ({type: SET_CURRENT_MOVIE, movie});
export const set_search_txt = (txt)=> ({type: SET_SEARCH_TXT, txt});
export const hydrate_store = () => {
    const retrieval = (window && window.localStorage)?
        JSON.parse(window.localStorage.getItem('netflixRouletteStore')):
        null;
    return {
        type: HYDRATE_STORE,
        retrievedStore: retrieval || {}
    };
};

export default {
    get_movies,
    sort_movies,
    set_current_movie,
    set_search_criterion,
    set_search_txt,
    hydrate_store
};