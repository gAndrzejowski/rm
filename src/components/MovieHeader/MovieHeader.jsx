// @flow
import React from 'react';
import { connect } from 'react-redux';
import styles from '../Heading/Heading.scss';
import PageTitle from '../PageTitle/PageTitle';
import BackToSearch from '../BackToSearch/BackToSearch';
import MoviePage from '../MoviePage/MoviePage';
import { setCurrentMovie } from '../../actions/creators';
import { getById } from '../../actions/async';
import type { MovieData, RouteMatch, Store } from '../../flowTypes';

type Props = {
  match: RouteMatch,
  selectedMovie: MovieData,
  backToSearch: () => void,
  getMovie: number => void,
}
export const MovieHeader = ({
  match: { params: { id } }, selectedMovie, backToSearch, getMovie,
}: Props) => {
  if (!selectedMovie || selectedMovie.id !== +id) {
    getMovie(id);
    return null;
  }
  return (
    <header>
      <div className={styles.top}>
        <PageTitle />
        <BackToSearch onClick={backToSearch} />
      </div>
      <MoviePage movie={selectedMovie} />
    </header>
  );
};

export const mapStateToProps = (state :Store) => ({
  selectedMovie: state.heading.selected,
});

export const mapDispatchToProps = (dispatch :any) => ({
  backToSearch: () :void => dispatch(setCurrentMovie(null)),
  getMovie: async (id :number) :Promise<any> => dispatch(await getById(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieHeader);
