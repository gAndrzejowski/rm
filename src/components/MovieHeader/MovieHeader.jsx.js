import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import styles from '../Heading/Heading.scss';
import PageTitle from '../PageTitle/PageTitle';
import BackToSearch from '../BackToSearch/BackToSearch.jsx';
import MoviePage from '../MoviePage/MoviePage';
import {set_current_movie} from "../../actions/creators";

export const MovieHeader = ({match: {params: {id}}, selectedMovie, backToSearch}) => (
    <header>
        <div className={styles.top}>
            <PageTitle/>
            <BackToSearch onClick={backToSearch}/>
        </div>
            <MoviePage movie={selectedMovie}/>
    </header>
);

MovieHeader.propTypes = {
    match: PropTypes.object,
    selectedMovie: PropTypes.object,
    backToSearch: PropTypes.func.isRequired
};

export const mapStateToProps = (state) => ({
    selectedMovie: state.heading.selectedMovie
});

export const mapDispatchToProps = (dispatch) => ({
    backToSearch: () => dispatch(set_current_movie(null))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieHeader);
