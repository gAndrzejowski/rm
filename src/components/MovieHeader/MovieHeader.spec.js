import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {MovieHeader, mapStateToProps, mapDispatchToProps} from "./MovieHeader.jsx";
import {SET_CURRENT_MOVIE} from '../../actions/names';

describe('MovieHeader component', () => {
    it('renders correctly', () => {
        const renderer = new ShallowRenderer();
        renderer.render(<MovieHeader match={{params:{id:3}}} selectedMovie={{}} backToSearch={jest.fn()} getMovie={jest.fn()} />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
    })
});

describe('store connection functions', () => {
   it('mapStateToProps gets selected movie data from state', () => {
       const fakeState = {
           heading: {
               selected: 'a movie'
           }
       };
       expect(mapStateToProps(fakeState)).toEqual({
           selectedMovie: 'a movie'
       });
   });
    it('mapDispatchToProps creates a backToSearch function that dispatches a SET_CURRENT_MOVIE action with null param', () => {
        const fakeDispatch = jest.fn();
        const props = mapDispatchToProps(fakeDispatch);
        expect(props.backToSearch).toBeInstanceOf(Function);
        props.backToSearch();
        expect(fakeDispatch).toHaveBeenCalledTimes(1);
        expect(fakeDispatch.mock.calls[0][0]).toEqual({
            type: SET_CURRENT_MOVIE,
            movie: null
        })
    })
});