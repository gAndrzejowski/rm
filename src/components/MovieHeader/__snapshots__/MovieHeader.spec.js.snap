// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieHeader component renders correctly 1`] = `
<header>
  <div
    className="top"
  >
    <PageTitle />
    <BackToSearch
      onClick={[MockFunction]}
    />
  </div>
  <MoviePage
    movie={Object {}}
  />
</header>
`;
